py/htm/encoders/eye.py

```diff
@@ -0,0 +1,588 @@
+            print("Magno Sparsity %g"%(np.count_nonzero(magno) / np.product(magno.shape)))
+        roi = self.make_roi_pretty()
+        cv2.imshow('Region Of Interest', roi)
+        cv2.imshow('Parvocellular', self.parvo[:,:,::-1])
+        cv2.imshow('Magnocellular', self.magno)
+        cv2.waitKey(1)
+
+    def input_space_sample_points(self, npoints):
+        """
+        Returns a sampling of coordinates which the eye is currently looking at.
+        Use the result to determine the actual label of the image in the area
+        where the eye is looking.
+        """
+        # Find the retina's radius in the image.
+        r = int(round(self.scale * self.retina_diameter / 2))
+        # Shrink the retina's radius so that sample points are nearer the fovea.
+        # Also shrink radius B/C this does not account for the diagonal
+        # distance, just the manhattan distance.
+        r = r * 2/3
+        # Generate points.
+        coords = np.random.random_integers(-r, r, size=(npoints, 2))
+        # Add this position offset.
+        coords += np.array(np.rint(self.position), dtype=np.int).reshape(1, 2)
+        return coords
+
+    def reset(self):
+        self.retina.clearBuffers()
+
+
+class EyeSensorSampler:
+    """
+    Samples eyesensor.rgb, the eye's view.
+
+    Attribute samples is list of RGB numpy arrays.
+    """
+    def __init__(self, eyesensor, sample_period, number_of_samples=30):
+        """
+        This draws its samples directly from the output of eyesensor.view() by
+        wrapping the method.
+        """
+        self.sensor         = sensor = eyesensor
+        self.sensor_compute = sensor.compute
+        self.sensor.compute = self.compute
+        self.age          = 0
+        self.samples      = []
+        number_of_samples = min(number_of_samples, sample_period)   # Don't die.
+        self.schedule     = random.sample(range(sample_period), number_of_samples)
+        self.schedule.sort(reverse=True)
+
+    def compute(self, *args, **kw_args):
+        """Wrapper around eyesensor.view which takes samples"""
+        retval = self.sensor_compute(*args, **kw_args)
+        if self.schedule and self.age == self.schedule[-1]:
+            self.schedule.pop()
+            roi = self.sensor.make_roi_pretty(self.sensor.roi)
+            self.samples.append(roi)
+        self.age += 1
+        return retval
+
+    def view_samples(self, show=True):
+        """Displays the samples."""
+        if not self.samples:
+            return  # Nothing to show...
+        plt.figure("Sample views")
+        num = len(self.samples)
+        rows = math.floor(num ** .5)
+        cols = math.ceil(num / rows)
+        for idx, img in enumerate(self.samples):
+            plt.subplot(rows, cols, idx+1)
+            plt.imshow(img[:,:,::-1], interpolation='nearest')
+        if show:
+            plt.show()
+
+
+# TODO: Consider splitting motor controls and motor sensory into different
+# classes...
```

Comment:
:+1:

Reply:
This code was part of a failed experiment and should be removed from this PR.  I do not fully understand how the brain controls the eye.

Instead the eye-encoder can just accept the coordinates, scale, and rotation of where the eye should look.

Reply:
I removed the motor control code.